File: bin/format_sniffles_vcf.pl
#!/usr/bin/perl
use strict;

my $usage = "Usage:   perl $0 <in.sniffles.vcf> <out_prefix>\n";
$usage   .= "Contact: Li Fang (fangli\@grandomics.com)\n";
$usage   .= "Version: 0.4.0\n";
die $usage if (@ARGV < 2);

my $min_length = 50;        # min sv length for DEL, DUP and INV
my $max_length = 10000000;   # max sv length for DEL, DUP and INV
my $min_read_supp = 2;
my $in = shift(@ARGV);
my $out_prefix = shift(@ARGV);
my $out1 = "$out_prefix.INS.bed";
my $out2 = "$out_prefix.DEL.bed";
my $out3 = "$out_prefix.INV.bed";
my $out4 = "$out_prefix.TRA.bedpe";   # not bed format
my $out5 = "$out_prefix.DUP.bed";

open (IN, $in) or die $!;
open (INS, "> $out1") or die $!;
open (DEL, "> $out2") or die $!;
open (INV, "> $out3") or die $!;
open (TRA, "> $out4") or die $!;
open (DUP, "> $out5") or die $!;

print STDERR "Format $in ... Settings: min length = $min_length, max length = $max_length\n";

my $header1 = "#Chr\tStart\tEnd\tRef\tAlt\tType\tLength\tSV_caller\tInfo\n";

print INS  $header1;
print DEL  $header1; 
print INV  $header1; 
print DUP  $header1; 

my %hash = ("1" => 1, "2" => 1, "3" => 1, "4" => 1, "5" => 1, "6" => 1, "7" => 1, "8" => 1, "9" => 1, "10" => 1, "11" => 1, "12" => 1, "13" => 1, "14" => 1, "15" => 1, "16" => 1, "17" => 1, "18" => 1, "19" => 1, "20" => 1, "21" => 1, "22" => 1, "X" => 1, "Y" => 1, "MT" => 1, "chr1" => 1, "chr2" => 1, "chr3" => 1, "chr4" => 1, "chr5" => 1, "chr6" => 1, "chr7" => 1, "chr8" => 1, "chr9" => 1, "chr10" => 1, "chr11" => 1, "chr12" => 1, "chr13" => 1, "chr14" => 1, "chr15" => 1, "chr16" => 1, "chr17" => 1, "chr18" => 1, "chr19" => 1, "chr20" => 1, "chr21" => 1, "chr22" => 1, "chrX" => 1, "chrY" => 1, "chrMT" => 1);

while (my $line = <IN>)
{
	chomp $line;
	(my $chr, my $start, my $id, my $ref, my $type, my $qual, my $filter, my $info, my $format, my $sample) = split("\t", $line);

	my $index = index($info, "CHR2=");
	my $chr2  = substr($info, $index+length("CHR2="));
	my @b = split(";", $chr2);
	my $chr2 = $b[0];

	my $index = index($info, "END=");
	my $end = substr($info, $index+length("END="));
	my @b = split(";", $end);
	my $end = $b[0];

	my $index = index($info, "SVLEN=");
	my $sv_length = substr($info, $index+length("SVLEN="));
	my @b = split(";", $sv_length);
	my $sv_length = $b[0];

	my $index = index($info, "RE=");
	my $rd_supp = substr($info, $index+length("RE="));
	my @b = split(";", $rd_supp);
	my $rd_supp = $b[0];
	next if ($rd_supp < $min_read_supp );

	if ($type =~ /TRA/){
		$type = "TRA";
		if (exists ($hash{$chr}) and exists ($hash{$chr2})){
			my $start1 = $start;
			my $end1 = $start1+1;
			my $start2 = $end;
			my $end2 = $start2+1;
			my $item = "$chr\t$start1\t$end1\t$chr2\t$start2\t$end2\tcomments:sv_type=$type;sv_caller=Sniffles;numReads=$rd_supp;$filter;$info;$format;$sample\n";
			print TRA $item;
		}
		next;
	}

	if ($type =~ /INS/){
		$type = "INS";
	}elsif($type =~ /DEL/){
		$type = "DEL";
	}elsif($type =~ /DUP/){
		$type = "DUP";
	}elsif($type =~ /INV/){
		$type = "INV";
	}

	$start = $start - 1; 

	if ($start > $end){
		print STDERR "Notice: start position ($start) is smaller than end ($end) ! Switch start and end.\n";
		my $temp = $start;
		$start = $end;
		$end = $temp;
	} # switch position if start < end;

	my $item = "$chr\t$start\t$end\t0\t0\tcomments:sv_type=$type;sv_length=$sv_length;sv_caller=Sniffles;numReads=$rd_supp;$filter;$info;$format;$sample\n";

######filter results######

	next if (!exists ($hash{$chr})); 
	next if ($sv_length < $min_length ); # filter sv length
	next if ($sv_length > $max_length ); # filter sv length
	next if ($end - $start > 1000 and $type eq "INS");     # fuzzy breakpoint of INS

######output results######

	if ($type eq "INS") {
		print INS $item;
	}elsif ($type eq "DEL") {
		print DEL $item;
	}elsif ($type eq "DUP") {
		print DUP $item;
	}elsif ($type eq "INV") {
		print INV $item;
	}
}

close IN;
close INS;
close DEL;
close INV;
close TRA;
close DUP;

File: bin/format_sniffles_vcf.py
#!/usr/bin/env python

import os
import sys

tab  = '\t'
endl = '\n'
arg = sys.argv[1:]

usage = 'python ' + __file__ + ' ' + '<in.vcf> <out_prefix> <min_read_supp>'
argc  = 3

class SVCall:

    def __init__(self, attr_list):

        self.chr1, self.pos1, self.id, self.ref, self.alt, self.qual, self.filter, self.info, self.format, self.genotype_info = attr_list

        self.pos1      = int(self.pos1)

        info_list      = self.info.split(';')
        self.chr2      = ''
        self.pos2      = -1
        self.std1      = -1
        self.std2      = -1
        self.kurt1     = -1000000
        self.kurt2     = -1000000
        self.svtype    = '.'
        self.supp_type = '.'
        self.sv_len    = 0 
        self.reported_sv_len = 0 
        self.strands   = ''
        self.read_supp = 0

        for info in info_list:

            if info == 'IMPRECISE': 
                self.precise = 'False'
                continue
            if info == 'PRECISE':
                self.precise = 'True'
                continue

            info = info.split('=')
            
            if len(info) != 2: continue
            key, value = info
            if key == 'CHR2':
                self.chr2 = value
            elif key == 'END':
                self.pos2 = int(value)
            elif key == 'STD_quant_start':
                self.std1 = float(value)
            elif key == 'STD_quant_stop':
                self.std2 = float(value)
            elif key == 'Kurtosis_quant_start':
                self.kurt1 = float(value)
            elif key == 'Kurtosis_quant_stop':
                self.kurt2 = float(value)
            elif key == 'SVTYPE':
                self.svtype = value
            elif key == 'SUPTYPE':
                self.supp_type = value
            elif key == 'SVLEN':
                self.reported_sv_len = int(value)
            elif key == 'STRANDS':
                self.strands = value
            elif key == 'RE':
                self.read_supp = int(value)
            elif key == 'RNAMES':
                self.readnames = value

        if self.chr1 == self.chr2:
            self.sv_len = abs(self.reported_sv_len)
        else:
            self.sv_len = -1
         
        a = list()
        if '[' in self.alt:
            a = self.alt.split('[')
        elif ']' in self.alt:
            a = self.alt.split(']')

        if len(a) == 3:
            bk2 = a[1]
            self.chr2, self.pos2 = a[1].split(':')
            self.pos2 = int(self.pos2)
        
        self.genotype  = './.'
        self.ref_reads = '.'
        self.alt_reads = '.'

        genotype_info = self.genotype_info.split(':')
        if len(genotype_info) == 3:
            self.genotype, self.ref_reads, self.alt_reads = genotype_info

        self.vaf = 'NA'
        if self.ref_reads != '.' and self.alt_reads != '.':
            ref_reads = float(self.ref_reads)
            alt_reads = float(self.alt_reads)
            vaf = alt_reads / (alt_reads + ref_reads) * 100
            self.vaf = '%.1f%%' % (vaf)

        
    # sv_type, sv_len, sv_id, filter, read_supp, 
    def output(self):

        if self.chr1 == self.chr2:
            outstring = '%s\t%d\t%d\t' % (self.chr1, self.pos1-1, self.pos2)
        else:
            outstring = '%s\t%d\t%d\t%s\t%d\t%d\t' % (self.chr1, self.pos1-1, self.pos1, self.chr2, self.pos2-1, self.pos2)

        outstring += '%s\t%d\tID=%s\t%s\t%d\t%s\t' % (self.svtype, self.sv_len, self.id, self.filter, self.read_supp, self.vaf)

        outstring += 'genotype=%s;ref_reads=%s;alt_reads=%s;supp_type=%s;precise=%s;std1=%.4f;std2=%.4f;kurt1=%.4f;kurt2=%.4f;strand=%s;\t' % (self.genotype, self.ref_reads, self.alt_reads, self.supp_type, self.precise, self.std1, self.std2, self.kurt1, self.kurt2, self.strands)

        outstring += 'ref=%s\talt=%s\t' % (self.ref, self.alt)
        outstring += '%s' % (self.readnames)

        return outstring

def main():

    if len(arg) < argc:
        print usage
        sys.exit()

    in_vcf_file = arg.pop(0)
    out_prefix  = arg.pop(0)
    min_read_supp = int(arg.pop(0))
      

    primary_chr_name = set()
    for i in range(1, 22):
        primary_chr_name.add(str(i))
        primary_chr_name.add('chr' + str(i))

    primary_chr_name.add('X')
    primary_chr_name.add('Y')
    primary_chr_name.add('M')
    primary_chr_name.add('MT')
    primary_chr_name.add('chrX')
    primary_chr_name.add('chrY')
    primary_chr_name.add('chrM')
    primary_chr_name.add('chrMT')

    max_sv_len = int(2e7)
    in_vcf_fp = open(in_vcf_file, 'r')

    out_ins_file = out_prefix + '.ins.bed'
    out_del_file = out_prefix + '.del.bed'
    out_dup_file = out_prefix + '.dup.bed'
    out_inv_file = out_prefix + '.inv.bed'
    out_tra_file = out_prefix + '.tra.bedpe'
    out_complex_file = out_prefix + '.complex.bed'

    out_ins_fp = open(out_ins_file, 'w')
    out_del_fp = open(out_del_file, 'w')
    out_dup_fp = open(out_dup_file, 'w')
    out_inv_fp = open(out_inv_file, 'w')
    out_tra_fp = open(out_tra_file, 'w')
    out_complex_fp = open(out_complex_file, 'w')

    while 1:
        line = in_vcf_fp.readline()
        if not line: break
        if line[0] == '#': continue
        line = line.strip().split(tab)
        sv = SVCall(line)
        if (sv.chr1 not in primary_chr_name) or (sv.chr2 not in primary_chr_name): continue
        
        ### filtering calls ###
        if sv.sv_len > max_sv_len: continue 
        if sv.read_supp < min_read_supp: continue
        if sv.chr1 != sv.chr2:
            out_tra_fp.write(sv.output() + endl)
        elif sv.svtype == 'INS':
            out_ins_fp.write(sv.output() + endl)
        elif sv.svtype == 'DEL':
            out_del_fp.write(sv.output() + endl)
        elif sv.svtype == 'DUP':
            out_dup_fp.write(sv.output() + endl)
        elif sv.svtype == 'INV':
            out_inv_fp.write(sv.output() + endl)
        else:
            out_complex_fp.write(sv.output() + endl)


    in_vcf_fp.close()
    out_ins_fp.close()
    out_del_fp.close()
    out_dup_fp.close()
    out_inv_fp.close()
    out_tra_fp.close()
    out_complex_fp.close()




if __name__ == '__main__':
    main()
